feat(VirusScanSchedulerService): Manual Scheduler service for deleting the attachment.

ravi110336 opened the pull request:
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
    

- The manual scheduler service "**Delete Attachment from Local FileSystem**" will allow the user to delete the attachment which has been uploaded to local file system for virus scan which is more than 30 days older.

> * Which issue is this pull request belonging to and how is it solving it? (#1183  )
> * Did you add or update any new dependencies that are required for your change? No

Issue: #1183 

### Suggest Reviewer
> You can suggest reviewers here with an @mention.

### How To Test?
> How should these changes be tested by the reviewer?

-  Check if there is attachment stored to local file system in the directory **/opt/sw360tempattachments**.
-  If no attachment present in local FS then upload attachment and update the project,component or release that should create the attachment in local FS.
-  If the attachment which is present is more than 30 days older then it should delete that attachment by clicking the Manual scheduler service (**Attachment Deletion From Local FS**) from the Admin tab or by scheduling it through a scheduler service.
- The scheduler service user should be able to schedule it every day in the productive environment but it is also configurable.
- For scheduling:**Schedule Attachment Deletion From Local FS**, For Cancel: **Cancel Scheduled Attachment Deletion From Local FS**
- If there is some attachment which is modified today but there exists some additional attachment which is more than 30 days older in this scenario only the attachment with 30 days older should get deleted.
- Added the CVESearch service also in the manual scheduling.
- If in the directory all the attachment is more than 30 days older then it should delete the attachment along with the directory.
![image](https://user-images.githubusercontent.com/56516845/117613324-5bdde400-b184-11eb-8eb6-e5de7de371d5.png)


> Have you implemented any additional tests? No

### Checklist
Must:
- [X] All related issues are referenced in commit messages and in PR

Signed-off-by: ravi110336 kumar.ravindra@siemens.com

JaideepPalit commented:
(from discussion): Manual trigger is already there. Scheduled based trigger needs to be implemented which runs automatically after certain interval of time.

ravi110336 commented:
Added the scheduler service to schedule the task for deletion of attachment in local FS, along with the manual deletion, also added the CVESearch for Manual deletion.

JaideepPalit commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/DatabaseHandlerUtil.java:
```diff
@@ -795,4 +808,52 @@ public static void saveAttachmentInFileSystem(AttachmentConnector attachmentConn
+            deleteAttachmentAndDirectory(targetDir, thresholdDate);
+            status = RequestStatus.SUCCESS;
+        } catch (IOException e) {
+            log.warn("Unable to delete attachment. ", e);
```
Maybe it could be log.error since it returns `RequestStatus.FAILURE`

JaideepPalit commented on backend/src/src-schedule/src/main/java/org/eclipse/sw360/schedule/timer/ScheduleConstants.java:
```diff
@@ -36,7 +36,10 @@ private ScheduleConstants(){}
     public static final String CVESEARCH_OFFSET_DEFAULT  = 0 + "" ; // default 00:00 am, in seconds
     public static final String CVESEARCH_INTERVAL_DEFAULT  = (24*60*60)+"" ; // default 24h, in seconds
-
+    public static final String DELETE_ATTACHMENT_OFFSET_DEFAULT  = 0 + "" ; // default 00:00 am, in seconds
```
Maybe it could be directly written as `"0"`

JaideepPalit commented on backend/src/src-schedule/src/main/java/org/eclipse/sw360/schedule/timer/ScheduleConstants.java:
```diff
@@ -36,7 +36,10 @@ private ScheduleConstants(){}
     public static final String CVESEARCH_INTERVAL_DEFAULT  = (24*60*60)+"" ; // default 24h, in seconds
-
+    public static final String DELETE_ATTACHMENT_OFFSET_DEFAULT  = 0 + "" ; // default 00:00 am, in seconds
+    public static final String DELETE_ATTACHMENT_INTERVAL_DEFAULT  = (24*60*60)+"" ; // default 24h, in seconds
```
Indentation `(24*60*60)+""`

JaideepPalit commented on backend/src/src-schedule/src/main/java/org/eclipse/sw360/schedule/timer/ScheduleConstants.java:
```diff
@@ -69,8 +72,29 @@ private ScheduleConstants(){}
         }
 
+        if(! props.containsKey(DELETE_ATTACHMENT_OFFSET_PROPERTY_NAME)){
+            log.info("Property "+ DELETE_ATTACHMENT_OFFSET_PROPERTY_NAME + " not set. Using default value.");
```
Maybe log could be `debug` since it check for a property and uses default value if not present.
Indentation `"Property "+`
Same for Line 86-87

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/admin/ScheduleAdminPortlet.java:
```diff
@@ -133,4 +142,56 @@ public void unscheduleAllServices(ActionRequest request, ActionResponse response
+            RequestStatus requestStatus = new ThriftClients().makeCvesearchClient().update();
+            setSessionMessage(request, requestStatus, "Task", "perform");
+        } catch (TException e) {
+            log.error(e);
```
It would be good to have some Error message describing the error along with Exception stacktrace.
Same for line 184, 174, 162

JaideepPalit commented on frontend/sw360-portlet/src/main/resources/META-INF/resources/html/admin/scheduleAdmin/view.jsp:
```diff
@@ -70,4 +85,40 @@
+                </tr>
+                <tr>
+                    <th><liferay-ui:message key="next.synchronization" /></th>
+                    <td>${deleteAttachmentNextSync}</td>
```
Kindly use `<sw360:out>` tag to display text in jsp. It prevents xml error

JaideepPalit commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/DatabaseHandlerUtil.java:
```diff
@@ -9,16 +9,25 @@
+import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.net.MalformedURLException;
```
Kindly remove unused imports

JaideepPalit commented on frontend/sw360-portlet/src/main/resources/META-INF/resources/html/admin/scheduleAdmin/view.jsp:
```diff
@@ -31,6 +31,21 @@
+</portlet:actionURL>
+
+<portlet:actionURL var="triggerDeleteAttachmentURL" name="triggerDeleteAttachment">
+</portlet:actionURL>
```
Kindly remove duplicate `portlet:actionURL` 
Line 34-35
Line43-44

JaideepPalit commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/DatabaseHandlerUtil.java:
```diff
@@ -795,4 +808,52 @@ public static void saveAttachmentInFileSystem(AttachmentConnector attachmentConn
+                public FileVisitResult postVisitDirectory(Path directory, IOException ioException) throws IOException
+                {
+                    DirectoryStream<Path> stream = Files.newDirectoryStream(directory);
+                    boolean isFolderEmpty =  !stream.iterator().hasNext();
```
Kindly check Indentation

JaideepPalit commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/DatabaseHandlerUtil.java:
```diff
@@ -795,4 +808,52 @@ public static void saveAttachmentInFileSystem(AttachmentConnector attachmentConn
\ No newline at end of file
+
+    public static RequestStatus deleteOldAttachmentFromFileSystem() {
+        String targetDir = ATTACHMENT_STORE_FILE_SYSTEM_LOCATION;
```
Maybe it is not needed to create a separate variable instead `ATTACHMENT_STORE_FILE_SYSTEM_LOCATION` could be used directly.
`String targetDir = ATTACHMENT_STORE_FILE_SYSTEM_LOCATION;`

JaideepPalit reviewed (changes requested):
Reviewed the PR.
Kindly have a look at some review comments

ravi110336 commented:
@JaideepPalit ,Thanks for the review, Incorporated the changes as per the comments, Please check.

JaideepPalit reviewed (approved):
Code looks fine to me.
Tested attachment deletion from local file system successfully.
